clippy_lints/src/matches.rs

```diff
@@ -664,6 +686,50 @@ fn check_match_as_ref(cx: &LateContext<'_, '_>, ex: &Expr<'_>, arms: &[Arm<'_>],
     }
 }
 
+fn check_wild_in_or_pats(cx: &LateContext<'_, '_>, ex: &Expr<'_>, arms: &[Arm<'_>]) {
+    let mut is_non_exhaustive_enum = false;
+    let ty = cx.tables.expr_ty(ex);
+    if ty.is_enum() {
+        if let ty::Adt(def, _) = ty.kind {
+            if def.is_variant_list_non_exhaustive() {
+                is_non_exhaustive_enum = true;
+            }
+        }
+    }
+
+    for arm in arms {
+        if let PatKind::Or(ref fields) = arm.pat.kind {
+            // look for multiple fields in this arm that contains at least one Wild pattern
+            if fields.len() > 1 && fields.iter().any(is_wild) {
+                span_lint_and_then(
+                    cx,
+                    WILDCARD_IN_OR_PATTERNS,
+                    arm.pat.span,
+                    "wildcard pattern covers any other pattern as it will match anyway.",
+                    |db| {
+                        // handle case where a non exhaustive enum is being used
+                        if is_non_exhaustive_enum {
```

Comment:
I think the suggestion to handle `_` explicitly applies to both exhaustive and non-exhaustive types.

Building a suggestion for this is a little bit more complicated, since the `| _` has to be removed from the `Or` pattern first and then a new line has to be added for the `_ => unimplemented!()` case. But this suggestion will be always `MaybeIncorrect` /  `HasPlaceholders`, so I'd suggest to just use `db.span_help(..)` here.

Reply:
> I think the suggestion to handle _ explicitly applies to both exhaustive and non-exhaustive types.

But if we `span_help` for any types, we will miss good suggestions (i.e.: for Strings). Shouldn't we `span_help` only for enum cases ? Or maybe i'm wrong and your comment was only about enums ?

Reply:
I don't see the difference. 

With enums:

```rust
match e {
    E::A => {/*...*/},
    E::B | _ => {/*...*/}, // This should either be `E::B => {}, _ => {}` or `_ => {}`
}
```

With strings:
```rust
match s {
    "foo" => {/*...*/},
    "bar" | _ => {/*...*/}, // This should either be `"bar" => {}, _ => {}` or `_ => {}`
}
```

Reply:
I don't see the difference. `"bar" | _` is wrong in the same way as `E::A | _` is. In both cases, either `_` was meant or `"bar" => {}, _ => {}`/`E::A => {}, _ => {}`. So suggesting to split it up is a good suggestion in all cases (enums, strings, numbers, ...) IMO.

Always replacing it with just `_ => {}` may be the same semantically, but I don't think this is intended behavior by the user. So this suggestion is `MaybeIncorrect`